library/src/main/java/org/eclipse/uprotocol/ULink.java

```diff
@@ -0,0 +1,779 @@
+    private final UUri mResponseUri;
+    private final UBusClient mClient;
+    private final Executor mCallbackExecutor;
+    private final ServiceLifecycleListener mServiceLifecycleListener;
+
+    private final ConcurrentHashMap<UUID, CompletableFuture<UPayload>> mRequests = new ConcurrentHashMap<>();
+    private final Object mRegistrationLock = new Object();
+    @GuardedBy("mRegistrationLock")
+    private final Map<UUri, URpcListener> mRequestListeners = new HashMap<>();
+    @GuardedBy("mRegistrationLock")
+    private final Map<UUri, Set<UListener>> mListeners = new HashMap<>();
+    @GuardedBy("mRegistrationLock")
+    private boolean mRegistrationExpired;
+
+    private final String mTag;
+    private boolean mVerboseLoggable;
+
+    private final ConnectionCallback mConnectionCallback = new ConnectionCallback() {
+        @Override
+        public void onConnected() {
+            mCallbackExecutor.execute(() -> {
+                renewRegistration();
+                mServiceLifecycleListener.onLifecycleChanged(ULink.this, true);
+            });
+        }
+
+        @Override
+        public void onDisconnected() {
+            mCallbackExecutor.execute(() -> {
+                release();
+                mServiceLifecycleListener.onLifecycleChanged(ULink.this, false);
+            });
+        }
+
+        @Override
+        public void onConnectionInterrupted() {
+            mCallbackExecutor.execute(() -> {
+                setRegistrationExpired();
+                mServiceLifecycleListener.onLifecycleChanged(ULink.this, false);
+            });
+        }
+    };
+
+    private final UListener mListener = new UListener() {
+        @Override
+        public void onReceive(UUri topic, UPayload payload, UAttributes attributes) {
+            handleMessage(buildMessage(topic, payload, attributes));
+        }
+
+        @Override
+        public void onReceive(UMessage message) {
+            handleMessage(message);
+        }
+    };
+
+    /**
+     * Callback to notify the lifecycle of the uBus.
+     *
+     * <p>Access to the uBus should happen
+     * after {@link ServiceLifecycleListener#onLifecycleChanged(ULink, boolean)} call with
+     * {@code ready} set {@code true}.
+     */
+    public interface ServiceLifecycleListener {
+        /**
+         * The uBus has gone through status change.
+         *
+         * @param link  A {@link ULink} object that was originally associated with this
+         *              listener from {@link #create(Context, Handler, ServiceLifecycleListener)} call.
+         * @param ready When {@code true}, the uBus is ready and all accesses are ok.
+         *              Otherwise it has crashed or killed and will be restarted.
+         */
+        void onLifecycleChanged(@NonNull ULink link, boolean ready);
+    }
+
+    @VisibleForTesting
+    ULink(@NonNull Context context, @Nullable UEntity entity, @Nullable UBusClient client, @Nullable Executor executor,
```

Comment:
Documentation is missing for the constructor to explain the passed parameters.

Reply:
private constructor. Nobody will see it.